clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -377,6 +397,12 @@ else if (metadata.error() == Errors.TOPIC_AUTHORIZATION_FAILED)
                 log.debug("Updating last seen epoch for partition {} from {} to epoch {} from new metadata", tp, currentEpoch, newEpoch);
                 lastSeenLeaderEpochs.put(tp, newEpoch);
                 return Optional.of(partitionMetadata);
+            // If both topic IDs were valid and the topic ID changed, update the metadata
+            } else if (!topicId.equals(Uuid.ZERO_UUID) && oldTopicId != null && !topicId.equals(oldTopicId)) {
+                log.debug("Topic ID for partition {} changed from {} to {}, so this topic must have been recreated. " +
+                                "Using the newly updated metadata.", tp, oldTopicId, topicId);
```

Comment:
Instead of "Using the newly updated metadata," maybe we can say this:
> Resetting the last seen epoch to {}.

Reply:
So part of the reason why I say this is that this method decides whether we use or ignore the new metadata to update the cache in general. But I can change to the epoch message.